components/txn_types/src/lock.rs

```diff
@@ -55,15 +56,57 @@ impl LockType {
     }
 }
 
+/// Used in `Lock` to indicate the kind of transaction which the lock is part of.
+/// Note that this is not the kind of lock - a pessimistic lock may only be used
+/// in a pessimistic transaction, but a pessimistic transaction may include other
+/// kinds of locks.
+#[derive(new, PartialEq, Copy, Clone, Debug)]
+pub enum TransactionKind {
+    Optimistic,
+    /// The field is the commit timestamp of data being read in this transaction.
+    Pessimistic(NonZeroTimeStamp),
+}
+
+impl TransactionKind {
+    fn write_bytes(self, b: &mut Vec<u8>) {
+        if let TransactionKind::Pessimistic(ts) = self {
+            b.push(FOR_UPDATE_TS_PREFIX);
+            b.encode_u64(ts.into_inner()).unwrap();
+        }
+    }
+
+    pub fn is_pessimistic(self) -> bool {
+        match self {
+            TransactionKind::Optimistic => false,
+            TransactionKind::Pessimistic(_) => true,
+        }
+    }
+
+    pub fn will_expire_before(self, ts: NonZeroTimeStamp) -> bool {
```

Comment:
Personally I think this function is a bit confusing because `expire` is ambiguous. Could you explain it in comments that it actually means whether this _pessimistic_ lock can be removed or replaced by another pessimistic lock.

Reply:
And the given `ts` is just to avoid removing or overwriting a newer pessimistic lock by an old request. I think it's not `expire`... (I'm trying to understand...)